### PR TITLE
Improve JSONEachRow reading by ignoring the keys case

### PR DESCRIPTION
**Pull request by KevinyhZou**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Allow matching column names in a case insensitive manner when reading json files (`input_format_json_case_insensitive_column_matching`).

So we introduce a config to ignore the case problem in such situation.
 

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

---
### Modify your CI run:
**NOTE:** If your merge the PR with modified CI you **MUST KNOW** what you are doing
**NOTE:** Set desired options before CI starts or re-push after updates

#### Run only:
- [ ]  Integration tests
- [ ]  Integration tests (arm64)
- [ ]  Stateless tests (release)
- [ ]  Stateless tests (asan)
- [ ]  Stateful tests (release)
- [ ]  Stateful tests (asan)
- [ ]  No sanitizers
- [ ]  Tests with analyzer
- [ ]  Fast tests
- [ ]  Only package_debug build
- [ ]  Add your CI variant description here

#### CI options:
- [ ]  do not test (only style check)
- [ ]  disable merge-commit (no merge from master before tests)
- [ ]  disable CI cache (job reuse)

#### Only specified batches in multi-batch jobs:
- [ ]  1
- [ ]  2
- [ ]  3
- [ ]  4

### CODE REVIEW

**Comment by robot-clickhouse** [comment]
*This is an automated comment for commit 63012430a460c940f5cdcd1142350bbe1d5c8e99 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/61750/63012430a460c940f5cdcd1142350bbe1d5c8e99/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

File: src/Processors/Formats/Impl/JSONEachRowRowInputFormat.cpp [hunk 1]
```diff
@@ -45,6 +45,14 @@ JSONEachRowRowInputFormat::JSONEachRowRowInputFormat(
 {
     const auto & header = getPort().getHeader();
     name_map = header.getNamesToIndexesMap();
+    if (format_settings_.json.ignore_key_case)
+    {
+        for (auto it = name_map.begin(); it != name_map.end(); ++it)
+        {
+            StringRef key = it->first;
+            transformFieldNameToLowerCase(key);
```
Comment (divanik) [inline src/Processors/Formats/Impl/JSONEachRowRowInputFormat.cpp:53]: As it seems to me, you try to mutate key of a hash table while iterating on it. Usually mutating of a key breaks insides of an iterator of a data structure

File: src/Processors/Formats/Impl/JSONEachRowRowInputFormat.h [hunk 2]
```diff
@@ -55,7 +55,13 @@ class JSONEachRowRowInputFormat : public IRowInputFormat
 
     virtual void readRowStart(MutableColumns &) {}
     virtual void skipRowStart() {}
-
+    void transformFieldNameToLowerCase(StringRef & field_name)
+    {
+        if (!field_name.data) return;
+        char * name_data = const_cast<char*>(field_name.data);
```
Comment (divanik) [inline src/Processors/Formats/Impl/JSONEachRowRowInputFormat.h:61]: I think that const_cast is a not good idea when you are trying to mutate sth that you don't own. I suggest the following: if key ignore_key_case is set to True, we will create map from the lower_case versions names to the actual names, and make more profound check while finding the column_name in our set

**Review by divanik — CHANGES_REQUESTED** [review]

**Review by divanik — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
I like the idea, but I'd prefer a few things differently:

1. It should apply to all data formats with names, such as CSVWithNames, Protobuf, Avro, Parquet, and even Native.
2. The modes should be match case, ignore case, and auto. The auto mode will search for a matching column case insensitive if the direct match isn't found. In contrast, the case insensitive mode should throw an exception if the data contain ambiguous names.

File: src/Core/Settings.h [hunk 3]
```diff
@@ -1046,6 +1046,7 @@ class IColumn;
     M(Bool, input_format_json_named_tuples_as_objects, true, "Deserialize named tuple columns as JSON objects", 0) \
     M(Bool, input_format_json_ignore_unknown_keys_in_named_tuple, true, "Ignore unknown keys in json object for named tuples", 0) \
     M(Bool, input_format_json_defaults_for_missing_elements_in_named_tuple, true, "Insert default value in named tuple element if it's missing in json object", 0) \
+    M(Bool, input_format_json_ignore_key_case, false, "Ignore json key case while read json field from string", 0) \
```
Comment (Avogar) [inline src/Core/Settings.h:1049]: Please, add this new setting to SettingsChangesHistory.h

File: tests/queries/0_stateless/02995_baseline_23_12_1.tsv [hunk 4]
```diff
@@ -378,6 +378,7 @@ input_format_json_read_objects_as_strings	1
 input_format_json_try_infer_named_tuples_from_objects	1
 input_format_json_try_infer_numbers_from_strings	0
 input_format_json_validate_types_from_metadata	1
+input_format_json_ignore_key_case	0
```
Comment (Avogar) [inline tests/queries/0_stateless/02995_baseline_23_12_1.tsv:381]: That's not right, this setting didn't exist in 23.12 version, please, don't change this file and add new setting to SettingsChangesHistory.h

**Comment by KevinyhZou** [comment]
done @Avogar

**Comment by KevinyhZou** [comment]
Could this pr be merged ? and support for other formats and modes can be later @divanik @alexey-milovidov

**Comment by alexey-milovidov** [comment]
It does not build, and the tests don't pass.

> Could this pr be merged ? and support for other formats and modes can be later

Not later. The code should be complete and consistent. The user experience should be flawless.

**Comment by KevinyhZou** [comment]
OK

**Comment by alexey-milovidov** [comment]
This is a good change - please resolve conflicts and merge.

**Comment by al13n321** [comment]
Test failures look unrelated:
 * 01158_zookeeper_log_long - flaky [1]
 * 02931_rewrite_sum_column_and_constant - https://github.com/ClickHouse/ClickHouse/pull/65800
 * performance comparison - shows lots of weird errors, it does that often these days, didn't investigate

[1] https://play.clickhouse.com/play?user=play#U0VMRUNUIGNoZWNrX3N0YXJ0X3RpbWUsIGNoZWNrX25hbWUsIHRlc3RfbmFtZSwgcHVsbF9yZXF1ZXN0X251bWJlciwgcmVwb3J0X3VybApGUk9NIGNoZWNrcwpXSEVSRSAxCiAgICBBTkQgY2hlY2tfc3RhcnRfdGltZSA+PSBub3coKSAtIElOVEVSVkFMIDcgZGF5CiAgICBBTkQgdGVzdF9zdGF0dXMgIT0gJ1NLSVBQRUQnCiAgICBBTkQgdGVzdF9zdGF0dXMgTElLRSAnRiUnCiAgICBBTkQgY2hlY2tfc3RhdHVzICE9ICdzdWNjZXNzJwogICAgYW5kIHRlc3RfbmFtZSBsaWtlICclMDExNThfem9va2VlcGVyX2xvZ19sb25nJScKT1JERVIgQlkgYWxs

File: src/Core/SettingsChangesHistory.h [hunk 5]
```diff
@@ -115,6 +115,7 @@ static const std::map<ClickHouseVersion, SettingsChangesHistory::SettingsChanges
               {"output_format_csv_serialize_tuple_into_separate_columns", true, true, "A new way of how interpret tuples in CSV format was added."},
               {"input_format_csv_deserialize_separate_columns_into_tuple", true, true, "A new way of how interpret tuples in CSV format was added."},
               {"input_format_csv_try_infer_strings_from_quoted_tuples", true, true, "A new way of how interpret tuples in CSV format was added."},
+              {"input_format_json_ignore_key_case", false, false, "Ignore json key case while read json field from string."},
```
Comment (Algunenano) [inline src/Core/SettingsChangesHistory.h:118]: This should be 24.7

**Comment by alexey-milovidov** [comment]
It is implemented not like we needed. Let's revert.

File: src/Processors/Formats/Impl/JSONEachRowRowInputFormat.h [hunk 6]
```diff
@@ -74,7 +80,8 @@ class JSONEachRowRowInputFormat : public IRowInputFormat
 
     /// Hash table match `field name -> position in the block`. NOTE You can use perfect hash map.
     Block::NameMap name_map;
-
+    /// Hash table match `lower_case field name -> field name in the block`.
+    std::unordered_map<String, StringRef> lower_case_name_map;
```
Comment (alexey-milovidov) [inline src/Processors/Formats/Impl/JSONEachRowRowInputFormat.h:84]: This is very bad.

File: src/Processors/Formats/Impl/JSONEachRowRowInputFormat.h [hunk 7]
```diff
@@ -55,7 +55,13 @@ class JSONEachRowRowInputFormat : public IRowInputFormat
 
     virtual void readRowStart(MutableColumns &) {}
     virtual void skipRowStart() {}
-
+    String transformFieldNameToLowerCase(const StringRef & field_name)
+    {
+        String field_name_str = field_name.toString();
+        std::transform(field_name_str.begin(), field_name_str.end(), field_name_str.begin(),
+            [](unsigned char c) { return std::tolower(c); });
```
Comment (alexey-milovidov) [inline src/Processors/Formats/Impl/JSONEachRowRowInputFormat.h:62]: This is wrong.

**Comment by alexey-milovidov** [comment]
This will be reverted, because the review is not addressed: https://github.com/ClickHouse/ClickHouse/pull/61750#issuecomment-2019201049

**Comment by KevinyhZou** [comment]
would you open another pr to slove this problem? @alexey-milovidov

**Comment by alexey-milovidov** [comment]
We should do it, but I didn't plan it for the next days. It will be very helpful if you resumbit.

**Pull request merged**